### ISSUE TITLE
Allow direct pixel access in Image

### ISSUE THREAD

**Issue by Wavesonics** [godotengine/godot-proposals#868]
**Describe the project you are working on:**
Falling sand type of game:
https://en.wikipedia.org/wiki/Falling-sand_game

**Describe the problem or limitation you are having in your project:**
Currently access to the pixel data in `Image` requires you to go through accessors such as:
`get_pixel()` and `set_pixel()`
And you must use `Color` instead of treating them as raw bytes. This extra overhead is a problem for Falling Sand style games.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Allow GDNative and GDScript access to the Image pixel data array.
I believe `Image::ptr` is the pointer we need access to, and using `bytePixelStride` we can index the array manually.

Once modified, the dirty regions can be re-uploaded to the GPU texture using existing features.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
How to expose this to GDNative is the big question that I don't know how to accomplish. The normal data types such as PackedByteArray are all COW as far as I'm aware, and that really wouldn't work for performance critical applications that this is trying to enable.

So long story short, I don't know how it would work...

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
I think you could get around it using an Engine Plugin?

**Is there a reason why this should be core and not an add-on in the asset library?:**
Sharing Engine Plugins is hard.

**Comment by akien-mga** [comment]
Isn't `Image.get_data()` and the `Image.data` property what you're looking for?

**Comment by Wavesonics** [comment]
@akien-mga I could be wrong here, so please correct me if so:
The problem is that `data` is a `Vector`, which is a **COW** data structure internally, so modifying anything in it causes a full copy of the whole structure?

Inside `Image` is a pointer `ptr`, which could be used to index to and directly modify the pixel data without any data structure in your way. But I don't know how you'd expose that to the GDNative layer...

**Comment by KoBeWi** [comment]
I think you usually do such things in shader. Especially when you have lots of sand.

**Comment by Wavesonics** [comment]
@KoBeWi the problem with that is you need a cpu side representation in order to perform game play logic on the pixel data (think Worms style pixel perfect collision).

These types of games predate GPUs too.

**Comment by bojidar-bg** [comment]
There was some discussion that such low-level access to Image could be implemented as a separate GDNative interface, likewise for low-level Mesh access.

There are already such interfaces for networking, XR, and video decoding.

**Comment by Wavesonics** [comment]
@bojidar-bg ah cool, this is what I was fishing for, how would that work exactly? Do you have an example from those existing cases?

**Comment by bojidar-bg** [comment]
For the interface adding network extensibility see:
* The C interface definition: https://github.com/godotengine/godot/blob/master/modules/gdnative/include/net/godot_net.h
* The same expressed as JSON for languages not supporting the C syntax: https://github.com/godotengine/godot/blob/master/modules/gdnative/gdnative_api.json#L7188
* The binding for setting the struct of user-provided function pointers on the class allowing extensions: https://github.com/godotengine/godot/blob/master/modules/gdnative/net/stream_peer_gdnative.cpp#L74

**Comment by Zylann** [comment]
#5342 is slightly different though. It is about updating an `ImageTexture`, while this proposal is for raw access to `Image` pixel data.

**Comment by viktor-ferenczi** [comment]
> @KoBeWi the problem with that is you need a cpu side representation in order to perform game play logic on the pixel data (think Worms style pixel perfect collision).

Facing the same problem here. Falling back to the ugly unsafe workaround of casting away the `const` from the pointer returned by `get_data().ptr()`. I really don't like to do this, but there is no alternative...
